IPSec Client Documentation (#715)
128technology/docs: docs/plugin_ipsec_client.md

The IPsec plugin setup has the following key parts to the configuration.
* `ipsec-profile` describing the mechanism with which to connect to the server.
* `ipsec-client` represent the remote endpoints or server with which the ipsec client communicates.
* `ipsec-client-settings` configure universal settings for all conductor-managed routers.
* `service-route`'s to route the traffic through the tunnels

### Profiles

> The `router > ipsec-profile`'s are reusable IPsec settings that can be used across multiple nodes in a router and multiple IPsec endpoint `remote`s. The examples below shows two examples of ipsec profiles that mutually exclusive one using pre-shared keys, and one using certificate based authentication. The examples below show two IPSec profiles that are mutually exclusive.

Reviewer: ```suggestion
The `router > ipsec-profile`'s are reusable IPsec settings that can be used across multiple nodes in a router and multiple IPsec endpoint `remote`s. The examples below shows two examples of ipsec profiles that are mutually exclusive; one using pre-shared keys, and one using certificate based authentication. The examples below show two IPSec profiles that are mutually exclusive.
```

The IPsec plugin setup has the following key parts to the configuration.
* `ipsec-profile` describing the mechanism with which to connect to the server.
* `ipsec-client` represent the remote endpoints or server with which the ipsec client communicates.
* `ipsec-client-settings` configure universal settings for all conductor-managed routers.
* `service-route`'s to route the traffic through the tunnels

### Profiles

> The `router > ipsec-profile`'s are reusable IPsec settings that can be used across multiple nodes in a router and multiple IPsec endpoint `remote`s. The examples below shows two examples of ipsec profiles that mutually exclusive one using pre-shared keys, and one using certificate based authentication.

Reviewer 2: `The examples below show two IPSec profiles that are mutually exclusive; ...`

Author: Added

:::note

> This plugin can only connect to IPsec endpoints that support pre-shared key or X.509 certificate.

Reviewer 2: maybe change `endpoints` to `servers`

### Profiles

> The `router > ipsec-profile`'s are reusable IPsec settings that can be used across multiple nodes in a router and multiple IPsec endpoint `remote`s. The examples below shows two examples of ipsec profiles that are mutually exclusive; one using pre-shared keys, and one using certificate based authentication. The examples below show two IPSec profiles that are mutually exclusive.

Reviewer 3: The last sentence is a repeat and should be removed.
"The examples below show two IPSec profiles that are mutually exclusive."

Author: Previous comments are misunderstood. Fixed.

            common-criteria-mode true
            private-key rem1-private-key
                name rem1-private-key
                content (removed)
            exit

            private-key rem2-private-key
                name rem2-private-key
                content (removed)
            exit
    exit
exit
```

#### Private Key

##### Version History

| Release  | Modification                         |
| -------- | ------------------------------------ |
| 3.7.0    | `ipsec-client-settings > private-key` introduced |

The `private-key` allows the users to configure private keys to be used for IPSec X.508 certificate type.

```
config
    authority
        ipsec-client-settings
            private-key rem1-private-key
                name rem1-private-key
                content (removed)
            exit
        exit
    exit
exit
```

| Config            | Type      | Description                                    |
| --------          | -----     | -------------------                            |
| name              | string    | The name of the the private key.               |
| content           | string    | Private key to be used for X.509 certificate.  |

:::warning

> The `private-key` is used for pkc12 certification creation which will be used for tunnel authentication. Wrongly configured private key may prevent IPSec tunnel from establishing successfully. 

Reviewer: ```suggestion
The `private-key` is used to create the pkc12 certificate for tunnel authentication. A wrongly configured private key may prevent an IPSec tunnel from being established successfully. 
```

Once enabled, the records will allow the IPsec controller to perform additional functions such as detecting and remediating stuck egress tunnel sessions and reporting the name of the WAN interface being used for the tunnel.

> ### Configure X.509 Certificates Type For Tunnel Authentication

Reviewer 3: Looks like this header was not removed:
"### Configure X.509 Certificates Type For Tunnel Authentication"

Author: Fixed

Once enabled, the records will allow the IPsec controller to perform additional functions such as detecting and remediating stuck egress tunnel sessions and reporting the name of the WAN interface being used for the tunnel.

### Configure X.509 Certificates Type For Tunnel Authentication

> The user could enable X.509 certificate type for tunnel authenticate by configuring valid [`private-key`](#private-key), [`client-certificate`](config_command_guide.md#configure-authority-client-certificate) and [`trusted-ca-certificate`](config_command_guide.md#configure-authority-trusted-ca-certificate) then refer their key names to respective fields in [`ipsec-profile`](#profiles) section, with which a PKCS12 file will be generated. IPSec NSS database will be used to store the generated PKCS12 file and wiil be directly used for tunnel authentication. A public Libreswan document is refered [here](https://libreswan.org/wiki/HOWTO:_Using_NSS_with_libreswan). The plugin requires users to generate/acquire their private key, a CA certificate file and user certificate file signed by the CA certificate offline by utilities mentioned in Libreswan document or other reliable sources (openssl). The plugin will take over the configuration from `Importing third-party files into NSS` section listed in the Libreswan document.

Reviewer: ```suggestion
### Configure X.509 Certificate-type for Tunnel Authentication

The IPsec plugin requires users to generate/acquire their private key, a CA certificate file, and user certificate file. This must be signed by the CA certificate offline by utilities mentioned in Libreswan document (or other reliable sources such as openssl). Refer to the public [HOWTO:_Using_NSS_with_libreswan document](https://libreswan.org/wiki/HOWTO:_Using_NSS_with_libreswan) for additional information. Note that the IPsec plugin will take over the configuration mentioned in `Importing third-party files into NSS` in the Libreswan document. 

Use the following steps to create the X.509 certificate-type for Tunnel Authentication. 

1. Configure the [`private-key`](#private-key).
2. Configure the [`client-certificate`](config_command_guide.md#configure-authority-client-certificate).
3. Configure the [`trusted-ca-certificate`](config_command_guide.md#configure-authority-trusted-ca-certificate).
4. Enter the key names for each of these items in their respective fields in the [`ipsec-profile`](#profiles).

This information is used to generate the PKCS12 file. The IPsec NSS database stores the generated PKCS12 file for tunnel authentication. 
```

## Release Notes

### Release 3.7.0

**Release Date:** May 10, 2024

**Router Version** 128T-ipsec-2.5.0-3

#### New Features and Improvements

> - **/I95-51716** Common Criteria Certification - VPN Protection Profile

Reviewer 3: leading forward slash needs to be removed.

Author: done
